src/dlob/dlob_node.rs

```diff
@@ -0,0 +1,493 @@
+            Node::VAMMNode(_) => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct OrderNode {
+    pub order: Order,
+    pub user_account: Pubkey,
+    pub node_type: NodeType,
+}
+
+impl OrderNode {
+    pub fn new(node_type: NodeType, order: Order, user_account: Pubkey) -> Self {
+        Self {
+            order,
+            user_account,
+            node_type,
+        }
+    }
+}
+
+impl DLOBNode for OrderNode {
+    fn get_price(&self, oracle_price_data: OraclePriceData, slot: u64) -> u64 {
+        get_limit_price(&self.order, &oracle_price_data, slot, None)
+    }
+
+    fn is_vamm_node(&self) -> bool {
+        false
+    }
+
+    fn is_base_filled(&self) -> bool {
+        self.order.base_asset_amount_filled == self.order.base_asset_amount
+    }
+
+    fn get_sort_value(&self, order: &Order) -> Option<i128> {
+        match self.node_type {
+            NodeType::TakingLimit => Some(order.slot.into()),
+            NodeType::RestingLimit => Some(order.price.into()),
+            NodeType::FloatingLimit => Some(order.oracle_price_offset.into()),
+            NodeType::Market => Some(order.slot.into()),
+            NodeType::Trigger => Some(order.trigger_price.into()),
+            NodeType::VAMM => None,
+        }
+    }
+
+    fn get_order(&self) -> &Order {
+        &self.order
+    }
+
+    fn get_user_account(&self) -> Pubkey {
+        self.user_account
+    }
+
+    fn set_order(&mut self, order: Order) {
+        self.order = order;
+    }
+
+    fn get_node_type(&self) -> NodeType {
+        self.node_type
+    }
+
+    fn set_node_type(&mut self, node_type: NodeType) {
+        self.node_type = node_type;
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct VAMMNode {
+    pub order: Order,
+    pub price: u64,
+}
+
+impl VAMMNode {
+    pub fn new(order: Order, price: u64) -> Self {
+        Self { order, price }
+    }
+}
+
+impl DLOBNode for VAMMNode {
+    fn get_price(&self, _oracle_price_data: OraclePriceData, _slot: u64) -> u64 {
+        self.price
+    }
+
+    fn is_vamm_node(&self) -> bool {
+        true
+    }
+
+    fn is_base_filled(&self) -> bool {
+        false
+    }
+
+    fn get_sort_value(&self, _order: &Order) -> Option<i128> {
+        None
+    }
+
+    fn get_order(&self) -> &Order {
+        &self.order
+    }
+
+    fn get_user_account(&self) -> Pubkey {
+        unimplemented!()
+    }
+
+    fn set_order(&mut self, _order: Order) {
+        unimplemented!()
+    }
+
+    fn get_node_type(&self) -> NodeType {
+        NodeType::VAMM
+    }
+
+    fn set_node_type(&mut self, _node_type: NodeType) {
+        unimplemented!()
+    }
+}
+
+pub(crate) fn create_node(node_type: NodeType, order: Order, user_account: Pubkey) -> Node {
+    Node::new(node_type, order, user_account)
+}
+
+pub(crate) fn get_order_signature(order_id: u32, user_account: Pubkey) -> String {
+    format!("{}-{}", order_id, user_account.to_string())
```

Comment:
not sure about the string key, something like this with `Display`/to_string method may be better

```rust
#[derive(Eq, Hash, PartialEq)]
struct OrderKey {
    user: Pubkey,
    order_id: u32,
}
```

`pub type OrderKey = (Pubkey, u32);` could also work

Reply:
i like the string key, i think for simplicity it makes a lot of sense